docs: add release notes for 1.46 (#31875)
microsoft/playwright: docs/src/release-notes-js.md

import LiteYouTube from '@site/src/components/LiteYouTube';

## Version 1.46

### TLS Client Certificates

> TLS Client Authentication allows the server to request a client certificate and verify it.
> This release allows you to specify them for the browser and when using the [APIRequestContext] (API Testing).

Reviewer: ```suggestion
TLS Client Authentication is a web security measure where clients authenticate themselves to the server using a client certificate.
This release adds support for client certificates in Playwright tests.
```

The fact that the server "requests the client certificate" is an implementation detail that the changelog audience doesn't care about. I've added the fact that this is about web security so it's easier for folks to understand, who have never heard about "TLS" before.

Also, you're talking about `APIRequestContext` here, but also about a browser context below? I'm assuming this works for both, so i've removed the mention.

Author: web-security sounds scary, I'd leave it out for now.

There were two dedicated feature requests, hence mentioning both separately.

> TLS Client Authentication allows the server to request a client certificate and verify it.
> This release allows you to specify them for the browser and when using the [APIRequestContext] (API Testing).
> The following snippet will add a client certificate for `https://example.com` which gets used in the test runs:

Reviewer 2: ```suggestion
Playwright now allows to supply client-side certificates, so that server can verify them, as specified by TLS Client Authentication.

The following snippet sets up a client certificate for `https://example.com`:
```

```ts
import { defineConfig } from '@playwright/test';

export default defineConfig({
  // ...
  use: {
    clientCertificates: [{
      origin: 'https://example.com',
      certPath: './cert.pem',
      keyPath: './key.pem',
      passphrase: 'mysecretpassword',
    }],
  },
  // ...
});
```

> Its also possible to manually specify them per-project ([project "use section"](./api/class-testproject#test-project-use)) or when creating a new context or APIRequestContext manually via [`method: Browser.newContext`] and [`method: APIRequest.newContext`].

Reviewer: ```suggestion
It's also possible to specify client certificates per-project ([project "use section"](./api/class-testproject#test-project-use)) or when creating a new context or APIRequestContext manually via [`method: Browser.newContext`] and [`method: APIRequest.newContext`].
```

I don't see what's manual about the per-project `use`, so i've removed that word. Also, replacing `them` with the word it refers to makes this easier to parse when skimming the text rapidly. (general rule for pronouns in technical writing - if it the repetition isn't yet painful, then it's probably a good idea to not use pronous and just repeat)

Author: Superseded by Dima's comment.

> Its also possible to manually specify them per-project ([project "use section"](./api/class-testproject#test-project-use)) or when creating a new context or APIRequestContext manually via [`method: Browser.newContext`] and [`method: APIRequest.newContext`].

Reviewer 2: ```suggestion
You can also provide client certificates to a particular [test project](./api/class-testproject#test-project-use) or as a parameter of [`method: Browser.newContext`] and [`method: APIRequest.newContext`].
```

### Component Testing: New `router` fixture

> This release introduces a `router` fixture to intercept and handle network requests.

Reviewer 2: ```suggestion
This release introduces an experimental `router` fixture to intercept and handle network requests.
```

> This release introduces a `router` fixture to intercept and handle network requests.

Reviewer: This currently sounds like the release introduces the ability to intercept network requests. My understanding is that we're not adding an ability, but we're making it *easier* API-wise. Would this capture that better?
```suggestion
This release introduces a `router` fixture that's a new API to intercept and handle network requests.
```

Reviewer: Only now did I realize that this is specific to component testing. How do you feel about this instead?

```diff
- This release introduces a `router` fixture to intercept and handle network requests.
+ This release introduces a `router` fixture to intercept and handle network requests in component tests.
```

There are two ways to use the router fixture:

- Call `router.route(url, handler)` that behaves similarly to [`method: Page.route`].
- Call `router.use(handlers)` and pass [MSW library](https://mswjs.io) request handlers to it.

Here is an example of reusing your existing MSW handlers in the test.

```ts
import { handlers } from '@src/mocks/handlers';

test.beforeEach(async ({ router }) => {
  // install common handlers before each test

>   await router.use(...handlers);

Reviewer: As a reader who knows about MSW, this reads fine. If the reader doesn't know about MSW though, this isn't super interesting. Would it make sense to instead show the entire enchilada, and put some MSW code into here? That way the reader would learn about an interesting tool we like.

});

test('example test', async ({ mount }) => {
  // test as usual, your handlers are active
  // ...
});
```

> For more information, see the [component testing docs](./test-components#handling-network-requests).

Reviewer 2: ```suggestion
This fixture is only available in [component tests](./test-components#handling-network-requests).
```

### Test runner

> - New CLI option `--only-changed=<sha>` to only run test files that have been changed between working tree and "ref".

Reviewer 2: ```suggestion
- New CLI option `--only-changed` to only run test files that have been changed since the last commit or from a specific git "ref".
```

> - Allow [boxing for fixtures](./test-fixtures#box-fixtures) to minimize the fixture exposure to the reporters UI

Reviewer 2: ```suggestion
- New option to [box a fixture](./test-fixtures#box-fixtures) to minimize the fixture exposure in test reports and error messages.
- New option to provide a [custom fixture title](./test-fixtures#custom-fixture-title) to be used in test reports and error messages.
```

### Miscellaneous

- Show request method/status in the network details tab in Trace Viewer

> - Allow hiding Route actions like `route.continue` in Trace Viewer

Reviewer 2: I think we can make a section with UI mode / Trace Viewer improvements, given a large list of updates there.

> - Support for `maxRetries` option in [`method: APIRequestContext.fetch`] which retries on `ECONNRESET`

Reviewer 2: ```suggestion
- New `maxRetries` option in [`method: APIRequestContext.fetch`] which retries on the `ECONNRESET` error.
```

> - Support for `maxRetries` option in [`method: APIRequestContext.fetch`] which retries on `ECONNRESET`

Reviewer: ```suggestion
- Support for `maxRetries` option in [`method: APIRequestContext.fetch`] which retries on the `ECONNRESET` network error
```

a little public service for people who have no idea what `ECONNRESET` is
